server/integration/testsuite/src/test/java/org/infinispan/server/test/partitionhandling/PartitionHandlingIT.java

```diff
@@ -0,0 +1,252 @@
+
+      assertThrowsHotRodClientException(() -> allowReadsCache3.put(server3OwnedKey, "value3"));
+      assertEquals("value", allowReadsCache3.get(server3OwnedKey));
+
+      healCluster();
+   }
+
+   @Test
+   public void testAllowReadWrites() throws InterruptedException {
+      RemoteCacheManager cacheManager = ITestUtils.createCacheManager(server2);
+      RemoteCache<Object, Object> cache = cacheManager.getCache(ALLOW_READ_WRITES_CACHE);
+
+      String server1OwnedKey = getKeyOwnedByNode(server1Address, cacheManager, cache, 1);
+      String server3OwnedKey = getKeyOwnedByNode(server3Address, cacheManager, cache, 1);
+
+      cache.put(server1OwnedKey, "value");
+      cache.put(server3OwnedKey, "value");
+
+      partitionCluster();
+
+      RemoteCache<Object, Object> allowReadWritesCache1 = ITestUtils.createCacheManager(server1).getCache(ALLOW_READ_WRITES_CACHE);
+      RemoteCache<Object, Object> allowReadWritesCache3 = ITestUtils.createCacheManager(server3).getCache(ALLOW_READ_WRITES_CACHE);
+
+      assertEquals("value", allowReadWritesCache1.get(server1OwnedKey));
+      allowReadWritesCache1.put(server1OwnedKey, "value1");
+      assertEquals("value1", allowReadWritesCache1.get(server1OwnedKey));
+      allowReadWritesCache1.put(server3OwnedKey, "value1");
+      assertEquals("value1", allowReadWritesCache1.get(server3OwnedKey));
+
+      assertEquals("value", allowReadWritesCache3.get(server3OwnedKey));
+      allowReadWritesCache3.put(server3OwnedKey, "value2");
+      assertEquals("value2", allowReadWritesCache3.get(server3OwnedKey));
+      allowReadWritesCache3.put(server1OwnedKey, "value2");
+      assertEquals("value2", allowReadWritesCache3.get(server1OwnedKey));
+
+      healCluster();
+   }
+
+   private void partitionCluster() {
+      partitionHandlingController.partitionCluster(new HashSet<>(Arrays.asList("node0", "node1")),
+                                                   new HashSet<>(Arrays.asList("node2")));
+   }
+
+   private void healCluster() {
+      partitionHandlingController.heal();
+   }
+
+   private void assertThrowsHotRodClientException(Runnable r) {
+      try {
+         r.run();
+         fail();
+      } catch (HotRodClientException hce){
+         // OK
+      }
+   }
+
+   private String getKeyOwnedByNode(String serverAddress, RemoteCacheManager rcm, RemoteCache cache, int owners) {
+      SegmentConsistentHash hash = createHash(cache, NUM_SEGMENTS, owners);
+
+      int segmentOwnedByNode = -1;
+      for (Map.Entry<SocketAddress, Set<Integer>> socketAddressToSegments : cache.getCacheTopologyInfo().getSegmentsPerServer().entrySet()) {
+         if (socketAddressToSegments.getKey().toString().equals(serverAddress)) {
+            segmentOwnedByNode = socketAddressToSegments.getValue().stream().findFirst().get();
+         }
+      }
+
+      int i = 0;
+      while (hash.getSegment(marshall(rcm, "key" + i)) != segmentOwnedByNode) {
+         i++;
```

Comment:
This can actually enter an infinite loop, trying to figure out why.

Reply:
I think you should check that `segmentOwnedByNode != -1` before entering the loop. Optionally you could allow the key to map to any segment owned by `serverAddress`, not just the first one.